src/poetry/utils/env/python_manager.py

```diff
@@ -1,52 +1,88 @@
 from __future__ import annotations
 
 import shutil
-import subprocess
 import sys
 
 from functools import cached_property
 from pathlib import Path
 from typing import TYPE_CHECKING
+from typing import cast
+from typing import overload
+
+import findpython
+import packaging.version
 
 from cleo.io.null_io import NullIO
 from cleo.io.outputs.output import Verbosity
 from poetry.core.constraints.version import Version
-from poetry.core.constraints.version import parse_constraint
 
-from poetry.utils._compat import decode
 from poetry.utils.env.exceptions import NoCompatiblePythonVersionFoundError
-from poetry.utils.env.script_strings import GET_PYTHON_VERSION_ONELINER
 
 
 if TYPE_CHECKING:
+    from collections.abc import Iterable
+
     from cleo.io.io import IO
+    from typing_extensions import Self
 
     from poetry.config.config import Config
     from poetry.poetry import Poetry
 
 
+class ShutilWhichPythonProvider(findpython.BaseProvider):  # type: ignore[misc]
```

Comment:
Nice thing. How did you find out about this?

Reply:
By looking in the code heh.